packages/core/src/agent/MessageSender.ts

```diff
@@ -94,25 +112,40 @@ export class MessageSender {
 
     // Retrieve DIDComm services
     const allServices = this.transportService.findDidCommServices(connection)
-    const reachableServices = allServices.filter((s) => !isDidCommTransportQueue(s.serviceEndpoint))
+    let reachableServices = allServices.filter((s) => !isDidCommTransportQueue(s.serviceEndpoint))
+    if (options && options.preferredTransport) {
+      reachableServices = [
+        ...new Set(
+          reachableServices
+            .filter((s) => s.serviceEndpoint.split(':')[0] === options.preferredTransport)
+            .concat(reachableServices)
+        ),
+      ]
+    }
     const queueService = allServices.find((s) => isDidCommTransportQueue(s.serviceEndpoint))
 
     this.logger.debug(
       `Found ${allServices.length} services for message to connection '${connection.id}' (${connection.theirLabel})`
     )
 
-    if (!this.outboundTransporter) {
+    if (this.outboundTransporters.length === 0 && !queueService) {
       throw new AriesFrameworkError('Agent has no outbound transporter!')
     }
 
     // Loop trough all available services and try to send the message
     for await (const service of reachableServices) {
       this.logger.debug(`Sending outbound message to service:`, { service })
       try {
-        await this.outboundTransporter.sendMessage({
-          payload: packedMessage,
-          endpoint: service.serviceEndpoint,
-        })
+        const protocol = service.serviceEndpoint.split(':')[0]
+        for (const transport of this.outboundTransporters) {
+          if (transport.supportedSchemes.includes(protocol)) {
+            await transport.sendMessage({
+              payload: packedMessage,
+              endpoint: service.serviceEndpoint,
+            })
+            break
+          }
+        }
```

Comment:
I think we should extract this to a separate method as it is duplicated a few times:

```suggestion
        const transports = this.findTransportsForService(service)
        for (const transport of transports) {
            await transport.sendMessage({
              payload: packedMessage,
              endpoint: service.serviceEndpoint,
            })
            break
        }
```